plasTeX/ConfigManager.py

```diff
@@ -0,0 +1,181 @@
+        self.value = self.valueType()(string)
+
+    @property
+    def value(self) -> T:
+        return self._value
+
+    @value.setter
+    def value(self, value: T):
+        if type(value) != self.valueType():
+            raise TypeError("Encountered {} when expecting {}".format(type(value).__name__, self.valueType().__name__))
+
+        self._value = value
+
+class BooleanOption(Option[bool]):
+    def registerArgparse(self, group: ArgumentGroup):
+        enables = [x for x in self.options if x[0] != "!"]
+        disables = [x[1:] for x in self.options if x[0] == "!"]
+
+        group.add_argument(*enables, dest=self.name, help=self.description + formatDefault(self.value), action='store_true', default=None)
+        if len(disables) > 0:
+            group.add_argument(*disables, dest=self.name, help=self.description + formatDefault(self.value), action='store_false', default=None)
+
+class MultiStringOption(Option[List[str]]):
+    def registerArgparse(self, group: ArgumentGroup):
+        group.add_argument(*self.options, dest=self.name, type=str, nargs="*", help=self.description, action="append")
+
+    def setFromString(self, string: str):
+        self.value.extend(shlex.split(string))
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
+            for entry in value:
+                self._value.extend(entry)
+
+class DictOption(ABC, Option[Dict[str, T]]):
+    @classmethod
+    @abstractmethod
+    def entryFromString(cls, entry: str) -> T:
+        """This convers a string into T"""
+
+    def set(self, key: str, value: str):
+        self.value[key] = self.entryFromString(value)
+
+    def setFromString(self, string: str):
+        raise ValueError("Calling setFromString on a DictOption. The individual items should be inserted with set.")
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        entries = data.get(self.name)
+        if entries is not None:
+            for key, val in entries:
+                self.set(key, val)
+
+class ConfigSection():
+    def __init__(self, name: str, parent: 'ConfigManager'):
+        self.name = name
+        self.parent = parent
+        self.data = {} # type: Dict[str, Option]
+
+    def __getitem__(self, key: str) -> Any:
+        value = self.data[key].value
+        wrapper = InterpolationWrapper(self.parent)
+        if isinstance(value, str):
+            return value % wrapper
+        elif isinstance(value, list) and len(value) > 0 and isinstance(value[0], str):
+            return [x % wrapper for x in value]
+        else:
+            return value
+
+    def __setitem__(self, key: str, value: Any):
+        if isinstance(value, Option):
+            if key in self.data:
+                raise ValueError
+            self.data[key] = value
+        else:
+            self.data[key].value = value
+
+    def get(self, key: str, default: Optional[Any]=None) -> Optional[Any]:
```

Comment:
Could you save lines in this class by inheriting from `dict`?

Reply:
No, because we want to retrieve the value of the option instead of the option object itself.